SAP/spartacus
# fix: CXSPA-7474 Move error handling logic to ExpressJS error handlers

This PR contains an adjustment of our solution to forward raw errors to ExpressJS and only there (with small help from HttpErrorHandlingInterceptor) distinguish types of error to handle it properly.
Follow-up PR for PoC: https://github.com/SAP/spartacus/pull/18900

QA steps:

1) Tests in storefront app:
- call HTTP error that may be caught during server-side rendering e.g use HTTP proxy OR 
  - modify `pages` property in `default-cms-config.ts` to generate 'CmsPageNotFoundHttpErrorResponse`
  - modify `components` property in `default-cms-config.ts` to generate unknown server error
- run npm `run dev:ssr`
- open the application in the browser
- verify whether the app fell back to CSR with the proper status code:
 - `index.html` returned with status `404` if `pages` endpoint has been modified in `default-cms-config.ts`
 - `index.html` returned with status `500` if `components` endpoint has been modified in `default-cms-config.ts`
 
 2) Tests in customer's app:
 - build and deploy Spartacus packages locally:
```npx ts-node ./tools/schematics/testing.ts in the SPA root folder```
- generate fresh ng17 application:
```npx @angular/cli@17 new my-app --standalone=false --ssr=false --style=scss --routing=false```
- go to the project's directory and install Spartacus with SSR from local packages:
```npx @angular/cli@17 add @spartacus/schematics@latest --ssr```
- use HTTP proxy or modify config for CMS endpoints to cause an error when rendering:
```ts
provideConfig(<OccConfig>{
      backend: {
        occ: {
          endpoints: {
            // components: 'cms/components2', FOR STATUS 500
            pages: 'cms/pages2', // FOR STATUS 404
          },
        },
      },
    }),
```
- in `server.ts` inside `app` function use handlers provided by Spartacus:
```ts 
  const indexHtmlContent = readFileSync(indexHtml, 'utf-8');

  /** */

  server.use(defaultErrorResponseHandlers(indexHtmlContent));
```
- run `npm run build` and `npm run serve:ssr:my-app` to verify if server error responses have been handled as in test case 1
- create a custom handler for errors (you can use the below example):
  - ```npm i -d @types/ejs ejs```
  - handle-page-not-found.ts:
  ```ts
  import { ErrorRequestHandler } from 'express';
  import ejs from 'ejs';

  export const handlePageNotFound: ErrorRequestHandler = async (
    err,
    _req,
    res,
    next,
  ) => {
    if (err instanceof CmsPageNotFoundOutboundHttpError) {
      const html = await ejs.renderFile('./404.ejs', {
        title: 'OPS! Page not found',
        text: 'The page you are looking for does not exist',
      });
      res.status(404).send(html);
    } else {
      next(err);
    }
  };
  ```
  - `404.ejs`:
  ```
  <!DOCTYPE HTML>
  <html lang="en">
    <head>
      <meta charset="UTF-8" />
      <meta http-equiv="X-UA-Compatible" content="IE=edge" />
      <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <title>Page Not Found</title>
    </head>
    <body>
      <h1><%= title %></h1>
      <p><%= text %></p>
    </body>
  </html>
  ```
  - in `server.ts`, use handler inside `app` function:
  ```ts 
   server.use(handlePageNotFound)
  ```
- run `npm run build` and `npm run serve:ssr:my-app` to verify if the custom handler works

closes [CXSPA-7474](https://jira.tools.sap/browse/CXSPA-7474)
